vscode-dotty/src/extension.ts

```diff
@@ -54,11 +54,14 @@ export function activate(context: ExtensionContext) {
   // package.json does not work)
   vscode.languages.setLanguageConfiguration("scala", {
     "indentationRules": {
-      // Auto-indent when pressing enter on a line matching this regexp
-      "increaseIndentPattern": /(((\b(then|else|do|catch|finally|yield|match|while|try|for|if|case)\b)|\bif\s*\(.*\)|=|=>|<-|=>>)\s*$)/,
-      // Auto-unindent disabled, because it doesn't work well with all
-      // indentation styles
-      "decreaseIndentPattern": /^.$/
+      // Auto-indent when pressing enter on a line matching this regexp, in details:
+      // 1. If they're not preceded by `end`, auto-indent after `while`, `for`, `match`, `try`, `if`
+      // 2. Auto-indent after `if ...` as long as it doesn't match `if ... then ...`
+      // 3. Auto-indent after `then`, `else`, `do`, `catch`, `finally`, `yield`, `case`, `=`, `=>`, `<-`, `=>>`x
+      "increaseIndentPattern":
+        /(((?<!\bend\b\s*?)\b(if|while|for|match|try))|(\bif\s+(?!.*?\bthen\b.*?$)[^\s]*?)|(\b(then|else|do|catch|finally|yield|case))|=|=>|<-|=>>)\s*?$/,
+      // Only auto-unindent completed `end` folowed by `while`, `for`, `match`, `try`, `if`
+      "decreaseIndentPattern": /(\bend\b\s*)\b(if|while|for|match|try)$/
```

Comment:
This is missing a check that there's only spaces on the line before `end`, use `^\s*end` instead of `\bend`.
Also to get this into the next vscode-dotty release, I would have to rebase my branch to include this commit before the release commit.

Reply:
Maybe we should put this in the release. 

How pressing is the release? I also found a couple of annoying integration with `then`+enter and auto-completion.

Reply:
not pressing

>  I also found a couple of annoying integration with then+enter and auto-completion.

I saw that too, but I think that can only be fixed on the language server side, not the vscode extension side

Reply:
We could fix it by adding the keywords as template suggestions

Reply:
Though the language server could recommend more precise completion